spec/acceptance/realtime/auth_spec.rb

```diff
@@ -1031,5 +1031,184 @@ def disconnect_transport(connection)
+              stop_reactor
+            end
+            client.connect
+          end
+        end
+      end
+
+      context 'when JWT token expires' do
+        before do
+          stub_const 'Ably::Models::TokenDetails::TOKEN_EXPIRY_BUFFER', 0 # allow token to be used even if about to expire
+          stub_const 'Ably::Auth::TOKEN_DEFAULTS', Ably::Auth::TOKEN_DEFAULTS.merge(renew_token_buffer: 0) # Ensure tokens issued expire immediately after issue
+        end
+        let(:token_callback) do
+          lambda do |token_params|
+            # Ably in all environments other than production will send AUTH 5 seconds before expiry, so
+            # we generate a JWT that expires in 5s so that the window for Realtime to send has passed
+            tokenResponse = Faraday.get "#{auth_url}?keyName=#{key_name}&keySecret=#{key_secret}&expiresIn=5"
+            tokenResponse.body
+          end
+        end
+        let(:client_options) { default_options.merge(use_token_auth: true, auth_callback: token_callback) }
+
+        it 'a new token is requested via auth_callback and the client gets reconnected' do
+          client.connection.once(:connected) do
+            original_token = auth.current_token_details
+            original_conn_id = client.connection.id
+
+            client.connection.once(:disconnected) do |state_change|
+              expect(state_change.reason.code).to eql(40142)
+
+              client.connection.once(:connected) do
+                expect(original_token).to_not eql(auth.current_token_details)
+                expect(original_conn_id).to eql(client.connection.id)
+                stop_reactor
+              end
+            end
+          end
+        end
+      end
+
+      context 'when the JWT token request includes a client_id' do
+        let(:client_id) { random_str }
+        let(:auth_callback) do
+          lambda do |token_params|
+            Ably::Rest::Client.new(default_options).auth.request_token({ client_id: client_id }, { auth_url: auth_url, auth_params: auth_params }).token
+          end
+        end
+        let(:client_options) { default_options.merge(auto_connect: false, auth_callback: auth_callback) }
+
+        it 'the client_id is the same that was specified in the auth_callback that generated the JWT token' do
+          client.connection.once(:connected) do
+            expect(client.auth.client_id).to eql(client_id)
+            stop_reactor
+          end
+          client.connect
+        end
+      end
+
+      context 'when the JWT token request includes a subscribe-only capability' do
+        let(:client_id) { random_str }
+        let(:basic_capability) { JSON.dump(channel_name => ['subscribe']) }
+        let(:auth_callback) do
+          lambda do |token_params|
+            Ably::Rest::Client.new(default_options).auth.request_token({ capability: basic_capability }, { auth_url: auth_url, auth_params: auth_params }).token
```

Comment:
As above, whilst I think a test for Rest clients generating tokens from `auth_urls` with `JWT`s is useful, I don't think we should be creating a level of indirection here that makes understanding this test that much hardder.  Why not simply make an HTTP request to the echo server directly and return that in `auth_callback`?